File: app/src/main/java/com/atom/vpn/demo/fragment/ConnectWithDedicatedIPFragment.java
/*
 * Copyright (c) 2018 ATOM SDK Demo.
 * All rights reserved.
 */

package com.atom.vpn.demo.fragment;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.os.Handler;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.atom.core.exceptions.AtomAPIException;
import com.atom.core.exceptions.AtomException;
import com.atom.core.exceptions.AtomValidationException;
import com.atom.core.models.AtomConfiguration;
import com.atom.core.models.Protocol;
import com.atom.sdk.android.AtomManager;
import com.atom.sdk.android.ConnectionDetails;
import com.atom.sdk.android.Errors;
import com.atom.sdk.android.ProtocolName;
import com.atom.sdk.android.ProtocolType;
import com.atom.sdk.android.VPNCredentials;
import com.atom.sdk.android.VPNProperties;
import com.atom.sdk.android.VPNStateListener;
import com.atom.sdk.android.common.Common;
import com.atom.vpn.demo.AtomDemoApplicationController;
import com.atom.vpn.demo.R;
import com.atom.vpn.demo.activity.ConnectActivity;
import com.atom.vpn.demo.common.Constants;
import com.atom.vpn.demo.common.logger.Log;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

import static com.atom.vpn.demo.common.Utilities.changeButtonState;
import static com.atom.vpn.demo.common.Utilities.changeButtonText;


public class ConnectWithDedicatedIPFragment extends Fragment implements VPNStateListener {

    private static final String TAG = "ConnectWithDedicatedIPFragment";
    private EditText etDedicatedIP;
    //IKEV is only supported protocol for dedicated ip
    private Protocol supportedProtocol;
    private Button btnConnect;

    private String uuid,vpnUsername,vpnPassword;

    public ConnectWithDedicatedIPFragment() {
        // Required empty public constructor
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        Bundle extras  = this.getArguments();
        if(extras!=null) {

            if (extras.containsKey("uuid")) {
                uuid =  extras.getString("uuid");
            }

            if (extras.containsKey("vpnUsername")) {
                vpnUsername =  extras.getString("vpnUsername");
            }

            if (extras.containsKey("vpnPassword")) {
                vpnPassword =  extras.getString("vpnPassword");
            }

        }

        AtomManager.addVPNStateListener(this);

        if(getActivity()!=null)
        new Handler().postDelayed(() -> AtomDemoApplicationController.getInstance().getAtomManager().bindIKEVStateService(getActivity()),500);

    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_connect_with_dedicatedip, container, false);
    }

    @SuppressLint("SetTextI18n")
    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);


        etDedicatedIP =  view.findViewById(R.id.etDedicatedIP);

        EditText etProtocol = view.findViewById(R.id.etProtocol);
        etProtocol.setClickable(false);

        btnConnect = view.findViewById(R.id.btnConnect);
        changeButtonText(getActivity(),btnConnect);
        btnConnect.setOnClickListener(v -> {

            if (AtomDemoApplicationController.getInstance().getAtomManager() != null) {

                if (AtomDemoApplicationController.getInstance().getAtomManager().getCurrentVpnStatus(getActivity()).equalsIgnoreCase(AtomManager.VPNStatus.CONNECTED)) {
                    AtomDemoApplicationController.getInstance().getAtomManager().disconnect(getActivity());

                } else if (AtomDemoApplicationController.getInstance().getAtomManager().getCurrentVpnStatus(getActivity()).equalsIgnoreCase(AtomManager.VPNStatus.CONNECTING)) {
                    AtomDemoApplicationController.getInstance().getAtomManager().cancel(getActivity());
                } else {
                    if (!TextUtils.isEmpty(etDedicatedIP.getText().toString().trim())) {

                        ConnectActivity connectActivity = (ConnectActivity)getActivity();
                        if(connectActivity!=null)
                        connectActivity.logWrapper.clear();

                        etDedicatedIP.setError(null);

                        if (supportedProtocol != null) {
                            try {
                                VPNProperties.Builder vpnPropertiesBuilder;
                                // Dedicated Host here

                                    vpnPropertiesBuilder = new VPNProperties.Builder(
                                            etDedicatedIP.getText().toString(), supportedProtocol);


                                        if (!TextUtils.isEmpty(vpnUsername) && !TextUtils.isEmpty(vpnPassword)) {
                                            AtomDemoApplicationController.getInstance().getAtomManager().setVPNCredentials(new VPNCredentials(vpnUsername, vpnPassword));
                                        } else if (!TextUtils.isEmpty(uuid)) {
                                            AtomDemoApplicationController.getInstance().getAtomManager().setUUID(uuid);
                                        }

                                        btnConnect.setText("Cancel");

                                        AtomDemoApplicationController.getInstance().getAtomManager().connect(getActivity(), vpnPropertiesBuilder.build());



                            } catch (AtomValidationException e) {
                                e.printStackTrace();
                            }
                        }


                    } else {
                        etDedicatedIP.setError(Constants.HostRequired);
                    }
                }
            }

        });

        // Only Ikev2 Protocol is supported from ATOM SDK
        supportedProtocol = new Protocol();
        supportedProtocol.setName(ProtocolName.IKEV);
        supportedProtocol.setProtocol(ProtocolType.IKEV);

        if(supportedProtocol!=null)
        etProtocol.setText(supportedProtocol.getName());


        if (AtomDemoApplicationController.getInstance().getAtomManager() != null) {
            if (AtomDemoApplicationController.getInstance().getAtomManager().getCurrentVpnStatus(getActivity()).equalsIgnoreCase(AtomManager.VPNStatus.CONNECTED)) {
                changeButtonState(btnConnect, "Disconnect");
            }
        }

    }

    @Override
    public void onConnected() {

    }

    @Override
    public void onConnected(ConnectionDetails connectionDetails) {
        Log.d("connected",   "Connected");
        changeButtonState(btnConnect, "Disconnect");
    }

    @Override
    public void onConnecting() {

    }

    @Override
    public void onConnecting(VPNProperties vpnProperties, AtomConfiguration atomConfiguration) {

    }

    @Override
    public void onRedialing(AtomException atomException, ConnectionDetails connectionDetails) {
        Log.d(TAG, atomException.getMessage());
    }

    @Override
    public void onDialError(AtomException atomException, ConnectionDetails connectionDetails) {

        if(!AtomDemoApplicationController.getInstance().getAtomManager().getCurrentVpnStatus(getActivity()).equalsIgnoreCase(AtomManager.VPNStatus.CONNECTED)) {
            Log.d("onDialError", atomException.getMessage());
            if (atomException.getCode() != Errors._5039)
                changeButtonState(btnConnect, "Connect");
        }

        Log.d(TAG, atomException.getCode() + "");
        Log.d(TAG, atomException.getMessage() + "");

        if (atomException.getException() instanceof AtomAPIException) {
            AtomAPIException atomAPIException = (AtomAPIException) atomException.getException();
            Log.d(TAG, atomAPIException.getErrorMessage()  + " - " +atomAPIException.getCode());
        }

    }


    @Override
    public void onStateChange(String state) {

        Log.d(TAG,  state );

        if (state.equalsIgnoreCase(AtomManager.VPNStatus.CONNECTING) || state.equalsIgnoreCase(VPNState.RECONNECTING)) {
            changeButtonState(btnConnect, "Cancel");
        }
    }


    @Override
    public void onDisconnected(boolean isCancelled) {

    }

    @Override
    public void onDisconnected(ConnectionDetails connectionDetails) {
        if (connectionDetails.isCancelled()) {
            Log.d(TAG,   "Cancelled");
        } else {
            Log.d(TAG,   "Disconnected");
        }

        changeButtonState(btnConnect, "Connect");

    }

    @Override
    public void onUnableToAccessInternet(AtomException atomException, ConnectionDetails connectionDetails) {

    }

    @Override
    public void onPacketsTransmitted(String s, String s1, String s2, String s3) {

    }


    @Override
    public void onDestroyView() {
        super.onDestroyView();
        AtomManager.removeVPNStateListener(this);
        if (AtomDemoApplicationController.getInstance().getAtomManager() != null) {

            AtomDemoApplicationController.getInstance().getAtomManager().unBindIKEVStateService(getActivity());
        }
    }

}

File: app/src/main/java/com/atom/vpn/demo/fragment/ConnectWithDedicatedIPFragment.kt
/*
 * Copyright (c) 2018 ATOM SDK Demo.
 * All rights reserved.
 */

package com.atom.vpn.demo.fragment

import android.annotation.SuppressLint
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.EditText
import androidx.fragment.app.Fragment
import com.atom.core.exceptions.AtomAPIException
import com.atom.core.exceptions.AtomException
import com.atom.core.exceptions.AtomValidationException
import com.atom.core.models.AtomConfiguration
import com.atom.core.models.Protocol
import com.atom.sdk.android.AtomManager
import com.atom.sdk.android.ConnectionDetails
import com.atom.sdk.android.Errors
import com.atom.sdk.android.ProtocolName
import com.atom.sdk.android.ProtocolType
import com.atom.sdk.android.VPNCredentials
import com.atom.sdk.android.VPNProperties
import com.atom.sdk.android.VPNStateListener
import com.atom.vpn.demo.R
import com.atom.vpn.demo.activity.ConnectActivity
import com.atom.vpn.demo.atom
import com.atom.vpn.demo.common.Constants
import com.atom.vpn.demo.common.Utilities.changeButtonState
import com.atom.vpn.demo.common.Utilities.changeButtonText
import com.atom.vpn.demo.common.logger.Log

class ConnectWithDedicatedIPFragment : Fragment(), VPNStateListener {

    private lateinit var etDedicatedIP: EditText

    //IKEV is only supported protocol for dedicated ip
    private var supportedProtocol: Protocol? = null
    private var btnConnect: Button? = null

    private var uuid: String? = null
    private var vpnUsername: String? = null
    private var vpnPassword: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val extras = arguments
        if (extras != null) {
            if (extras.containsKey("uuid")) {
                uuid = extras.getString("uuid")
            }
            if (extras.containsKey("vpnUsername")) {
                vpnUsername = extras.getString("vpnUsername")
            }
            if (extras.containsKey("vpnPassword")) {
                vpnPassword = extras.getString("vpnPassword")
            }
        }

        AtomManager.addVPNStateListener(this)

        val act = activity
        if (act != null) {
            Handler(Looper.getMainLooper()).postDelayed(
                { atom?.bindIKEVStateService(act) },
                500
            )
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_connect_with_dedicatedip, container, false)
    }

    @SuppressLint("SetTextI18n")
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        etDedicatedIP = view.findViewById(R.id.etDedicatedIP)

        val etProtocol = view.findViewById<EditText>(R.id.etProtocol)
        etProtocol.isClickable = false

        btnConnect = view.findViewById(R.id.btnConnect)
        changeButtonText(requireContext(), btnConnect)
        btnConnect?.setOnClickListener {

            val manager = atom
            if (manager != null) {
                val ctx = requireContext()
                val status = manager.getCurrentVpnStatus(ctx)

                if (status.equals(AtomManager.VPNStatus.CONNECTED, ignoreCase = true)) {
                    manager.disconnect(ctx)
                } else if (status.equals(AtomManager.VPNStatus.CONNECTING, ignoreCase = true)) {
                    manager.cancel(ctx)
                } else {
                    if (etDedicatedIP.text.toString().trim().isNotEmpty()) {

                        (activity as? ConnectActivity)?.logWrapper?.clear()

                        etDedicatedIP.error = null

                        val protocol = supportedProtocol
                        if (protocol != null) {
                            try {
                                // Dedicated Host here
                                val vpnPropertiesBuilder = VPNProperties.Builder(
                                    etDedicatedIP.text.toString(), protocol
                                )

                                val user = vpnUsername
                                val pass = vpnPassword
                                val id = uuid
                                if (!user.isNullOrEmpty() && !pass.isNullOrEmpty()) {
                                    manager.setVPNCredentials(VPNCredentials(user, pass))
                                } else if (!id.isNullOrEmpty()) {
                                    @Suppress("DEPRECATION")
                                    manager.setUUID(id)
                                }

                                btnConnect?.text = "Cancel"

                                manager.connect(ctx, vpnPropertiesBuilder.build())

                            } catch (e: AtomValidationException) {
                                e.printStackTrace()
                            }
                        }

                    } else {
                        etDedicatedIP.error = Constants.HostRequired
                    }
                }
            }
        }

        // Only Ikev2 Protocol is supported from ATOM SDK
        val protocol = Protocol()
        protocol.name = ProtocolName.IKEV
        protocol.protocol = ProtocolType.IKEV
        supportedProtocol = protocol

        etProtocol.setText(protocol.name)

        val manager = atom
        if (manager != null &&
            manager.getCurrentVpnStatus(requireContext())
                .equals(AtomManager.VPNStatus.CONNECTED, ignoreCase = true)
        ) {
            changeButtonState(btnConnect, "Disconnect")
        }
    }

    // Legacy overload, deprecated by the SDK in favour of onConnected(ConnectionDetails).
    @Suppress("OVERRIDE_DEPRECATION")
    override fun onConnected() {
    }

    override fun onConnected(connectionDetails: ConnectionDetails) {
        Log.d("connected", "Connected")
        changeButtonState(btnConnect, "Disconnect")
    }

    // Legacy overload, deprecated by the SDK in favour of onConnecting(VPNProperties, ...).
    @Suppress("OVERRIDE_DEPRECATION")
    override fun onConnecting() {
    }

    override fun onConnecting(
        vpnProperties: VPNProperties,
        atomConfiguration: AtomConfiguration
    ) {
    }

    override fun onRedialing(
        atomException: AtomException,
        connectionDetails: ConnectionDetails
    ) {
        Log.d(TAG, atomException.message)
    }

    override fun onDialError(
        atomException: AtomException,
        connectionDetails: ConnectionDetails
    ) {

        val manager = atom
        val ctx = context
        if (manager != null && ctx != null &&
            !manager.getCurrentVpnStatus(ctx)
                .equals(AtomManager.VPNStatus.CONNECTED, ignoreCase = true)
        ) {
            Log.d("onDialError", atomException.message)
            if (atomException.code != Errors._5039) {
                changeButtonState(btnConnect, "Connect")
            }
        }

        Log.d(TAG, atomException.code.toString())
        Log.d(TAG, atomException.message)

        val inner = atomException.exception
        if (inner is AtomAPIException) {
            Log.d(TAG, "${inner.errorMessage} - ${inner.code}")
        }
    }

    override fun onStateChange(state: String) {

        Log.d(TAG, state)

        if (state.equals(AtomManager.VPNStatus.CONNECTING, ignoreCase = true) ||
            state.equals(VPNStateListener.VPNState.RECONNECTING, ignoreCase = true)
        ) {
            changeButtonState(btnConnect, "Cancel")
        }
    }

    override fun onPaused(e: AtomException, connectionDetails: ConnectionDetails) {
    }

    override fun onDisconnecting(connectionDetails: ConnectionDetails) {
    }

    // Legacy overload, deprecated by the SDK in favour of onDisconnected(ConnectionDetails).
    @Suppress("OVERRIDE_DEPRECATION")
    override fun onDisconnected(isCancelled: Boolean) {
    }

    override fun onDisconnected(connectionDetails: ConnectionDetails) {
        if (connectionDetails.isCancelled) {
            Log.d(TAG, "Cancelled")
        } else {
            Log.d(TAG, "Disconnected")
        }

        changeButtonState(btnConnect, "Connect")
    }

    override fun onUnableToAccessInternet(
        atomException: AtomException,
        connectionDetails: ConnectionDetails
    ) {
    }

    override fun onPacketsTransmitted(s: String, s1: String, s2: String, s3: String) {
    }

    override fun onDestroyView() {
        super.onDestroyView()
        AtomManager.removeVPNStateListener(this)
        val act = activity
        if (act != null) {
            atom?.unBindIKEVStateService(act)
        }
    }

    companion object {
        private const val TAG = "ConnectWithDedicatedIPFragment"
    }
}
